pynicotine/networkfilter.py

```diff
@@ -60,27 +60,33 @@ def _request_ip(self, user, action, list_type):
         core.queue.append(slskmessages.GetPeerAddress(user))
         return True
 
-    def _add_user_ip_to_list(self, user, list_type):
-        """ Add the current IP of a user to a list. """
+    def _add_user_ip_to_list(self, list_type, user, ip_address):
+        """ Add the current IP address and username of a user to a list """
 
         if list_type == "ban":
             ip_list = config.sections["server"]["ipblocklist"]
         else:
             ip_list = config.sections["server"]["ipignorelist"]
 
-        if self._request_ip(user, "add", list_type):
-            return None
+        if ip_address and not user:
+            # Try to get a username from currently active connections
+            user = self.get_known_username(ip_address) or ""
+
+        elif user and not ip_address:
+            # Try to get an address from currently active connections
+            ip_address = self.get_known_ip_address(user)
 
-        ip_address, _port = core.user_addresses[user]
+            if ip_address is None:
+                return self._request_ip(user, "add", list_type)
```

Comment:
Return value should be an IP address (or None), but you're returning a boolean value here.

Reply:
Yes, good point I wasn't really sure what to do in this case, handling of offline users in here certainly isn't ideal.

For now, we return a username, this at least avoids a boolean being expressed in the output for unknown/offline users, albeit at the expense of code readability.

Reply:
We can check if the return value `is_ip_address()`, this is only used in ban_user_ip() to close a banned connection. 

Beforehand, I don't think we even checked it wasn't None until the CloseConnectionIP() message to see if any of the conn_obj items were equal to None in the core (or having wildcard values containing `*` asktricks), so the return chain didn't really work properly anyway, and did result in some message pack error on certain inputs. That issue is solved now.

If you can think of a better way of doing this then feel free to change it or suggest something, this is the best I can come up with.

Reply:
> Return value should be an IP address (or None)

It's more desirable to press ahead with creating the filter with a dummy address, because then this allows it to be updated by the callback when the user comes on-line (or when reconnecting if the filter is added whilst offline).

No return is needed here at all @mathiascode